include/api/CDataFrameAnalysisConfigReader.h

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+#ifndef INCLUDED_ml_api_CDataFrameAnalysisConfigReader_h
+#define INCLUDED_ml_api_CDataFrameAnalysisConfigReader_h
+
+#include <core/CLogger.h>
+
+#include <api/ImportExport.h>
+
+#include <rapidjson/document.h>
+
+#include <map>
+#include <string>
+#include <type_traits>
+#include <vector>
+
+namespace ml {
+namespace api {
+
+//! \brief Reads and validates parameters for a data frame analysis.
+//!
+//! DESCRIPTION:\n
+//! This wraps up extracting parameter values from a JSON configuration object.
+//! It supports predefining a set of expected parameters. It is expected that
+//! there will be one static reader object per analysis type.
+//!
+//! The read method extracts all the parameters the JSON contains and returns
+//! a collection to get them by name.
+//!
+//! Expected usage is:
+//! \code
+//! static const CDataFrameAnalysisConfigReader reader{[] {
+//!     CDataFrameAnalysisConfigReader theReader;
+//!     theReader.addParameter("foo", CDataFrameAnalysisConfigReader::E_OptionalParameter);
+//!     theReader.addParameter("bar", CDataFrameAnalysisConfigReader::E_RequiredParameter);
+//! }()};
+//!
+//! auto parameters = reader.read(json);
+//! bool foo{parameters["foo"].fallback(true)};
```

Comment:
nit: what about "default" instead of "fallback"?

Reply:
I'd have preferred that, but it is a reserved word.